engine/src/main/java/org/hibernate/search/engine/cfg/impl/EngineConfigurationUtils.java

```diff
@@ -9,29 +9,46 @@
 import java.util.Optional;
 
 import org.hibernate.search.engine.cfg.BackendSettings;
-import org.hibernate.search.engine.cfg.ConfigurationPropertySource;
 import org.hibernate.search.engine.cfg.EngineSettings;
+import org.hibernate.search.engine.cfg.spi.ConfigurationScopeNamespace;
+import org.hibernate.search.engine.cfg.spi.ScopedConfigurationPropertySource;
+import org.hibernate.search.engine.environment.bean.BeanResolver;
 
 public final class EngineConfigurationUtils {
 
 	private EngineConfigurationUtils() {
 	}
 
-	public static ConfigurationPropertySourceExtractor extractorForBackend(Optional<String> backendNameOptional) {
+	public static ConfigurationPropertySourceExtractor extractorForBackend(BeanResolver beanResolver,
+			Optional<String> backendNameOptional) {
 		if ( !backendNameOptional.isPresent() ) {
-			return engineSource -> engineSource.withMask( EngineSettings.Radicals.BACKEND );
+			return engineSource -> engineSource
+					.withScope( beanResolver, ConfigurationScopeNamespace.BACKEND )
+					.withMask( EngineSettings.Radicals.BACKEND );
```

Comment:
I don't understand why you need to push the notion of scope into `ConfigurationPropertySource` and can't simply use something like this here:

```java
			return engineSource -> engineSource.withMask( EngineSettings.Radicals.BACKEND )
						.withOverride( configurationProvider.get( new ConfigurationScope(ConfigurationScopeNamespace.BACKEND, backendNameOptional ) );
```

Reply:
I still don't understand... can you please explain what's wrong with the example above?

Reply:
mmm so for this part:
```java
new ConfigurationScope(ConfigurationScopeNamespace.BACKEND, null )
```
we don't want to do that just inline since we'd be losing the scope sequence and the provider would only get the "backend", but what we want them to know is `global: backend: backend (some-backend-name): ....`  i.e. how did we get to the scope we are asking for. Let's say we are at the index scope and we have multiple  backends, per each backend we'd want to set a different index override - so we'd need to know from which backend we are coming from....

now as to why we can't simply apply `withFallback`. Let's look at a different example. Assuming we have a property that we want to set at all scopes to its own value. As soon as we apply the first fallback (global one) that key will have a value attached to it. And if the property is redefined for any other deeper scope -- that value is going to be ignored:
```
user source().
    withFallback(globalScopeSource(mykey=myvalue)
    .withFallback(backendScopeSource(mykey=myothervalue)
```
will always give us `myvalue` if we ask for a `source.get("mykey")`

so what we want to happen is this instead:
user source().
    .withFallback(backendScopeSource(mykey=myothervalue)
    withFallback(globalScopeSource(mykey=myvalue)
```
but when we are applying a new scope (or trying to add a new fallback) our property source is already modified and has some fallbacks attached to it. So we need to get way back to the `user source()`... but then we also have all those prefix/mask/override etc applied here and there...

Reply:
an alternative-alternative I've been thinking about was to add some sort of priority to the property... i.e. even if the fallback is applied at some later stage but has a higher priority the value from that fallback would "win". But that would mean that the get methods of property sources would need to go through the tree and compare these priorities... which seems to complicate things even more 😃

Reply:
> we don't want to do that just inline since we'd be losing the scope sequence and the provider would only get the "backend", but what we want them to know is global: backend: backend (some-backend-name): .... i.e. how did we get to the scope we are asking for. Let's say we are at the index scope and we have multiple backends, per each backend we'd want to set a different index override - so we'd need to know from which backend we are coming from....

Right, sure, we need a sequence of scopes. No argument there, my example was incomplete.

> Let's look at a different example. Assuming we have a property that we want to set at all scopes to its own value

We don't, though... ? Each scope has a different set of available properties. In places where properties can be applied to multiple scopes, I'd expect us to just apply defaults for each scope.

E.g. index properties can also be applied at the backend scope, to set defaults. So what we have currently (for the default backend) is something like this:

```java
userSource().withMask("hibernate.search.backend.indexes.<indexName>.")
        .withFallback(userSource().withMask("hibernate.search.backend."))
```

I'd have expected you to just do this:

```java
userSource().withMask("hibernate.search.backend.indexes.<indexName>.")
        .withFallback(<defaults from ConfigurationProvider for index <indexName>>)
        .withFallback(userSource().withMask("hibernate.search.backend."))
        .withFallback(<defaults from ConfigurationProvider for the default backend>)
```

And this just... doesn't require any change to `ConfigurationSource`?

Reply:
> E.g. index properties can also be applied at the backend scope, to set defaults. So what we have currently (for the default backend) is something like this:

Well, not really, we don't have a clean user source anywhere but at the very beginning. When we are extracting a property source for the index - we only have access to the backend source. At that point, the fallback is already applied to it. 
Now if we drag the unmodified user property source, we would need to apply the masks and fallbacks to it when we are about to extract properties. So that means either having the modifying operations saved somewhere or applying them as you showed in the example... but that seems error-prone to apply it like that...
In the second example, I think we need to have `.withFallback(<defaults from ConfigurationProvider for the global scope>)` at the end, too, I'd expect that the global scope can define any property (or more generally -- any scope for that matter can define a property that is "deeper"), even a backend/index specific one.
Anyway, if you think that we don't need this, a couple of questions before I make changes:
- where'd you save the current-scope-so-far?
- should we pass around and store in class fields the clean user source and apply the transformations in methods that are using the source?

Reply:
> Well, not really, we don't have a clean user source anywhere but at the very beginning. When we are extracting a property source for the index - we only have access to the backend source. At that point, the fallback is already applied to it.

Ok, but that just means we'd have this:

```java
var userSource = ...;
var backendSource = userSource.withMask("hibernate.search.backend.")
        .withFallback(<defaults from ConfigurationProvider for the default backend>);
var indexSource = backendSource().withMask("indexes.<indexName>.")
        .withFallback(<defaults from ConfigurationProvider for index <indexName>>);
```

This seems to still match what we need? The only downside is that defaults from the backend that use the prefix `indexes.<indexName>` would take precedence over defaults from the index, but, well... That seems unlikely. Whoever implements `ConfigurationProvider` just needs to be careful about not introducing conflicts.

> but that seems error-prone to apply it like that...

I'm not seeing how exactly this is error prone, could you please explain?

>  In the second example, I think we need to have .withFallback(<defaults from ConfigurationProvider for the global scope>) at the end, too, I'd expect that the global scope can define any property (or more generally -- any scope for that matter can define a property that is "deeper"), even a backend/index specific one.

I'd expect that to be applied much earlier, e.g.:

```java
var userSource = <something>.withFallback(<defaults from ConfigurationProvider for the global scope>);
var backendSource = userSource.withMask("hibernate.search.backend.")
        .withFallback(<defaults from ConfigurationProvider for the default backend>);
var indexSource = backendSource().withMask("indexes.<indexName>.")
        .withFallback(<defaults from ConfigurationProvider for index <indexName>>);
```

> Anyway, if you think that we don't need this

Depends what you mean by "this" xD

> where'd you save the current-scope-so-far?

I think in most (all?) cases we can rebuild it from the context. E.g. in `org.hibernate.search.engine.common.impl.IndexManagerBuildingStateHolder.BackendInitialBuildState#createIndexManagerBuildingState` we have access to `backendBuildContext.backendName()`.

>    should we pass around and store in class fields the clean user source and apply the transformations in methods that are using the source?

From what I can see you don't necessarily need the "clean" user source. Deriving the backend source from the "defaulted" global one, or the index source from the "defaulted" backend one, seems fine.

Reply:
Discussed through other channels, we agreed to go with the "dumb" solution I suggested, mainly because `ConfigurationProvider` is a (relatively obscure) SPI so we can reasonably require from implementors that they take care not to introduce conflicts between properties from e.g. the global scope and the index scope.